README.md
# EMBO - Empirical Bottleneck
[![License](https://img.shields.io/pypi/l/embo)](https://www.gnu.org/licenses/gpl-3.0.txt)
[![PyPI version](https://img.shields.io/pypi/v/embo.svg)](https://pypi.python.org/pypi/embo/)
[![Build status](https://img.shields.io/gitlab/pipeline/epiasini/embo)](https://gitlab.com/epiasini/embo/pipelines)

A Python package for working with the Information Bottleneck [[Tishby,
Pereira, Bialek 2001]](https://arxiv.org/abs/physics/0004057) and the
Deterministic (and Generalized) Information Bottleneck [[Strouse and
Schwab 2016]](https://arxiv.org/abs/1604.00268). Embo is especially
geared towards the analysis of concrete, finite-size data sets.

## Requirements

`embo` requires Python 3, `numpy>=1.7` and `scipy`.

## Installation
To install the latest release, run:
``` bash
pip install embo
```
(depending on your system, you may need to use `pip3` instead of `pip`
in the command above).

### Testing
(requires `setuptools`). If `embo` is already installed on your
system, look for the copy of the `test_embo.py` script installed
alongside the rest of the `embo` files and execute it. For example:

``` bash
python /usr/lib/python3.X/site-packages/embo/test/test_embo.py
```

## Usage

### The Information Bottleneck
We refer to [[Tishby, Pereira, Bialek
2001]](https://arxiv.org/abs/physics/0004057) and [[Strouse and Schwab
2016]](https://arxiv.org/abs/1604.00268) for a general introduction to
the Information Bottleneck. Briefly, if X and Y are two random
variables, we are interested in finding another random variable M
(called the "bottleneck" variable) that solves the following
optimisation problem:

min_{p(m|x)} = H(M) - α H(M|X) - β I(M:Y)

for any β>0 and 0≤α≤1, and where M is constrained to be independent on
Y conditional on X:

p(x,m,y) = p(x)p(m|x)p(y|x)

Intuitively, we want to find the stochastic mapping p(M|X) that
extracts from X as much information about Y as possible while
forgetting all irrelevant information. β is a free parameter that sets
the relative importance of forgetting irrelevant information versus
remembering useful information. α determines what notion of
"forgetting" we use: α=1 ("vanilla" bottleneck or IB) implies that we
want to minimise the mutual information I(M:X), α=0 (deterministic
bottleneck or DIB) that we want to make M a good *compression* of X by
minimising its entropy H(M), and intermediate values interpolate
between these two conditions.

Typically, one is interested in the curve described by I(M:Y) as a
function of I(M:X) or H(M) at the solution of the bottleneck problem
for a range of values of β. This curve gives the optimal tradeoff of
compression and prediction, telling us what is the minimum amount of
information one needs to know about X (or minimum amount of entropy
one needs to retain) to be able to predict Y to a certain accuracy, or
vice versa, what is the maximum accuracy one can have in predicting Y
given a certain amount of information about X.

### Using `embo`
Embo can solve the information bottleneck problem for discrete random
variables starting from a set of joint empirical observations. The
main point of entry to the package is the `InformationBottleneck`
class. In its constructor, `InformationBottleneck` takes as arguments an
array of observations for X and an (equally long) array of
observations for Y, together with other optional parameters (see the
docstring for details). In the most basic use case, users can call the
`get_bottleneck` method of an `InformationBottleneck` object, which will
assume α=1 and return a set of β values and the optimal values of
I(M:X), I(M:Y) and H(M) corresponding to those β. The optimal tradeoff
can then be visualised by plotting I(M:Y) vs I(M:Y).

For instance:

``` python
import numpy as np
from matplotlib import pyplot as plt
from embo import InformationBottleneck

# data sequences
x = np.array([0,0,0,1,0,1,0,1,0,1])
y = np.array([0,1,0,1,0,1,0,1,0,1])

# compute the IB bound from the data (vanilla IB; Tishby et al 2001)
I_x,I_y,H_m,β = InformationBottleneck(x,y).get_bottleneck()

# plot the IB bound
plt.plot(I_x,I_y)
```

Embo can also operate starting from a joint (X,Y) probability
distribution, encoded as a 2D array containing the probability
of each combination of states for X and Y.

``` python
# define joint probability mass function for a 2x2 joint pmf
pxy = np.array([[0.1, 0.4],[0.35, 0.15]]),

# compute IB
I_x,I_y,H_m,β = InformationBottleneck(pxy=pxy).get_bottleneck()

# plot I(M:Y) vs I(M:X)
plt.plot(I_x,I_y)
```

The deterministic and generalised bottleneck can be computed by
setting appropriately the parameter `alpha`:

``` python
# compute Deterministic Information Bottleneck (Strouse 2016)
I_x,I_y,H_m,β = InformationBottleneck(pxy=pxy, alpha=0).get_bottleneck()

# plot I(M:Y) vs H(M)
plt.plot(H_m,I_y)
```

### More examples
The `embo/examples` directory contains some Jupyter notebook that
should exemplify most of the package's functionality.

- [Basic-Example.ipynb](embo/examples/Basic-Example.ipynb): basics;
  how to compute and plot an IB bound.
- [Markov-Chains.ipynb](embo/examples/Markov-Chains.ipynb): using embo
  for *past-future bottleneck* type analyses on data from Markov
  chains.
- [Deterministic-Bottleneck.ipynb](embo/examples/Deterministic-Bottleneck.ipynb):
  Deterministic and Generalized Information Bottleneck. Here we
  reproduce a key figure from the Deterministic Bottleneck paper, and
  we explore the algorithm's behaviour as α changes from 0 to 1.
- [Compare-embo-dit.ipynb](embo/examples/Compare-embo-dit.ipynb): here
  we compare embo with [dit](https://pypi.org/project/dit) [[James et
  al 2018]](https://doi.org/10.21105/joss.00738). We compare the
  solutions found by the two packages on a set of simple IB problems
  (including a problem taken from dit's documentation), and we show
  that embo is orders of magnitude faster than dit.

### Further details
For more details, please consult the docstrings in
`InformationBottleneck`.

## Changelog
See the [CHANGELOG.md](CHANGELOG.md) file for a list of changes from
older versions.

## Authors
`embo` is maintained by Eugenio Piasini, Alexandre Filipowicz and
Jonathan Levine.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/), and this project
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2021-02-22
### Added
- Added support for Generalized Information Bottleneck (GIB;
  introduced by [Strouse and Schwab
  2016](https://arxiv.org/abs/1604.00268)). This includes the
  "vanilla" IB and the Deterministic IB as special cases.
- Improved performance by vectorizing some key sections of the code.
- Added support for computing IB/DIB/GIB directly from a table giving
  a joint probability distribution p(x,y) rather than empirical data.
- Added detailed example notebook for the Deterministic Information
  Bottleneck (`examples/Deterministic-Bottleneck.ipynb`).  This
  reproduces a key figure from the DIB paper and showcases how easy it
  now is to turn a knob and go from IB to DIB via various flavours of
  GIB.
- Added example notebook (`examples/Compare-embo-dit.ipynb`) with a
  comparison vs [dit](https://doi.org/10.21105/joss.00738), another
  information theory package that also implements the IB.  Here we
  show that the two packages give matching results on a few sample
  problems (except when dit crashes), and we compare embo and dit
  speed-wise. Embo is much faster.
- Added tests for GIB (actually just took most previous tests and made
  them to run over a range of IB types, from basic IB to DIB passing
  by some intermediate ones)
- Added tests to check that our internal implementation of KL
  divergence and entropy give exactly the same results as the
  corresponding functions from SciPy (they should, as we use the same
  low-level functions from `scipy.special`).
- Added tests on randomly-generated empirical data, sampled from a
  fixed probability distribution. This is better than what we had
  before because it makes it easier to catch corner cases.
- Added tests on randomly-generated probability distributions of
  various shapes and sizes. This is motivated along the same lines as
  the previous point; only it's the whole distribution that is random
  here, not (just) the data.
- Always run all tests with multiple versions of NumPy
  (1.18,1.19,1.20).
- Add license notices to files containing the core code.
- Added changelog file.
### Changed
- Renamed `EmpiricalBottleneck` class to `InformationBottleneck`, and
  `get_empirical_bottleneck` method to `get_bottleneck`.
- Made a small fix to the convergence criterion of the algorithm. This
  is nicer from a motivation standpoint (we monitor the loss function
  instead of the candidate pmf for changes), and also matches what is
  done in other papers such as Strouse.
- Changed initialization rule for the bottleneck "encoder" (p(m|x)),
  using a recipe inspired by what was done in the DIB paper. This was
  done mostly to aid in the comparison with the figure from that paper
  in the example notebooks, but it seems like a nice thing to have in
  general so I've switched to that by default. The previous
  initialization method is still used when the dimensionality of M is
  lower than that of X.
- Improved handling of corrupted input (NaNs, Infs, empty arrays)
- Switched to [tox](https://tox.readthedocs.io/en/latest/index.html)
  for managing unit tests, as using pure setuptools via `python
  setup.py test` is now discouraged.
- Removed usage of dtypes such as `np.int`, which were generating
  numpy deprecation warnings in NumPy 1.20.
- Removed some somewhat misleading/unnecessary references to
  past/future bottleneck in the docstrings and comments, where more
  general notions of IB were involved.


embo/embo.py
from __future__ import division
import numpy as np
from scipy.stats import entropy
from .utils import p_joint, mi_x1x2_c, compute_upper_bound
import multiprocessing as mp
np.seterr(divide='ignore', invalid='ignore')


def empirical_bottleneck(x,y,numuniquex=0,numuniquey=0,**kw):
    """ Compute an IB curve for two empirical sequences x and y - main function
    
    Arguments:
    x -- first empirical sequence (past)
    y -- second empirical sequence (future)
    
    Returns:
    i_p -- values of ipast for each value of beta
    i_f -- values of ifuture for each value of beta
    beta -- values of beta considered
    mixy -- mutual information between x and y (curve saturation point)
    hx -- entropy of x
    hy -- entropy of y
    """
    
    # Marginal, joint and conditional distributions required to calculate the IB
    pxy_j = p_joint(x,y)
    px = pxy_j.sum(axis=1)
    py = pxy_j.sum(axis=0)
    pyx_c = pxy_j.T / px
    
    #Calculate the information bottleneck for different values of beta
    i_p,i_f,beta,mixy,hx = IB(px,py,pyx_c,**kw)
    hy = entropy(py,base=2)
    
    # Return array of ipasts and ifutures for array of different values of beta - mixy should correspond to the saturation point
    return i_p,i_f,beta,mixy,hx,hy
    
def beta_iter(b,px,py,pyx_c,pm_size,restarts,iterations):
    """Function to run BA algorithm for individual values of beta
    
    Arguments:
    b -- value of beta on which to run algorithm
    px -- marginal probability distribution for the past (x)
    py -- marginal probability distribution for the future (y)
    pyx_c -- conditional distribution p(y|x)
    pm_size -- discrete size of the compression distribution
    restarts -- number of times the optimization procedure should be restarted (for each value of beta) from different random initial conditions
    iterations -- maximum number of iterations to use until convergence
    
    Returns:
    list with i_p and i_f, which correspond to ipast and ifuture values for each value of beta
    """
    candidates = []
    for r in range(restarts):
    
	    # Initialize distribution for bottleneck variable
	    pm = np.random.rand(pm_size)+1
	    pm /= pm.sum()                  
	    pym_c = np.random.rand(py.size,pm.size)+1 # Starting point for the algorithm
	    pym_c /= pym_c.sum(axis=0)
	    
	    # Iterate the BA algorithm
	    for i in range(iterations):
		    pmx_c, z = p_mx_c(pm,px,py,pyx_c,pym_c,b)
		    pm = p_m(pmx_c,px)
		    pym_c = p_ym_c(pm,px,py,pyx_c,pmx_c)
		    if i>0 and np.allclose(pmx_c,pmx_c_old,rtol=1e-3,atol=1e-3):
				# if the x->m mapping is not updating any more, we're at convergence and we can stop
			    break
		    pmx_c_old = pmx_c
	    candidates.append({'past_info' : mi_x1x2_c(pm, px, pmx_c),
						   'future_info' : mi_x1x2_c(py, pm, pym_c),
						   'functional' : -np.log2(np.inner(z,px))})
    # among the restarts, select the result that gives the minimum
    # value for the functional we're actually minimizing (eq 29 in
    # Tishby et al 2000).
    selected_candidate = min(candidates, key=lambda c: c['functional'])
    i_p = selected_candidate['past_info']
    i_f = selected_candidate['future_info']
    return [i_p,i_f,b]


def IB(px,py,pyx_c,maxbeta=5,numbeta=30,iterations=100,restarts=3,processes=1):
    """Compute an Information Bottleneck curve
    
    Arguments:
    px -- marginal probability distribution for the past
    py -- marginal distribution for the future
    maxbeta -- the maximum value of beta to use to compute the curve
    iterations -- number of iterations to use to for the curve to converge for each value of beta
    restarts -- number of times the optimization procedure should be restarted (for each value of beta) from different random initial conditions.
    processes -- number of cpu threads to run in parallel (default = 1)
    
    Returns:
    ips -- values of ipast for each beta considered
    ifs -- values of ifuture for each beta considered
    bs -- values of beta considered
    mixy -- mutual information between x and y (curve saturation point)
    hx -- entropy of x (maximum ipast value)
    """

    pm_size = px.size                      #Get compression size 
    bs = np.linspace(0.01,maxbeta,numbeta) #value of beta
    
    #Parallel computing of compression for desired beta values
    pool = mp.Pool(processes=processes)
    results = [pool.apply_async(beta_iter,args=(b,px,py,pyx_c,pm_size,restarts,iterations,)) for b in bs]
    pool.close()
    results = [p.get() for p in results]
    ips = [x[0] for x in results]
    ifs = [x[1] for x in results]
    
    #Values of beta may not be sorted appropriately, code below sorts ipast and ifuture according to their corresponding value of beta, and in correct order
    b_s = [x[2] for x in results]          
    ips = [x for _, x in sorted(zip(b_s,ips))]
    ifs = [x for _, x in sorted(zip(b_s,ifs))]
    
    # restrict the returned values to those that, at each value of
    # beta, actually increase (for Ipast) and do not decrease (for
    # Ifuture) the information with respect to the previous value of
    # beta. This is to avoid confounds from cases where the AB
    # algorithm gets stuck in a local minimum.
    ub, bs = compute_upper_bound(ips, ifs, bs)
    ips = np.squeeze(ub[:,0])
    ifs = np.squeeze(ub[:,1])
    
    # Return saturation point (mixy) and max horizontal axis (hx)
    mixy = mi_x1x2_c(py, px, pyx_c)
    hx = entropy(px,base=2)
    return ips, ifs, bs, mixy, hx


def p_mx_c(pm,px,py,pyx_c,pym_c,beta):
    """Update conditional distribution of bottleneck random variable given x.
    
    Arguments:
    pm -- marginal distribution p(M) - vector
    px -- marginal distribution p(X) - vector
    py -- marginal distribution p(Y) - vector
    pyx_c -- conditional distribution p(Y|X) - matrix
    pym_c -- conditional distribution p(Y|M) - matrix
    
    Returns:
    pmx_c -- conditional distribution p(M|X)
    z -- normalizing factor
    """
    
    pmx_c = np.zeros((pm.size,px.size)) # P(M|X) matrix to be returned
    for mi in range(pm.size):
        for xi in range(px.size):
            pmx_c[mi,xi] = pm[mi] * np.exp(-beta * entropy(pyx_c[:,xi], pym_c[:,mi], base=2))
    z = pmx_c.sum(axis=0)
    pmx_c /= z #Normalize 
    return pmx_c, z


def p_ym_c(pm,px,py,pyx_c,pmx_c):
    """Update conditional distribution of bottleneck variable given y.
    
    Arguments:
    pm -- Marginal distribution p(M)
    px -- marginal distribution p(X)
    pyx_c -- conditional distribution p(Y|X)
    pmx_c -- conditional distribution p(M|X)
    
    Returns:
    pym_c -- conditional distribution p(Y|M)
    """
    pym_c = np.zeros((py.size,pm.size))
    for yi in range(py.size):
        for mi in range(pm.size):
            for xi in range(px.size):
                pym_c[yi,mi] += (1./pm[mi])*pyx_c[yi,xi]*pmx_c[mi,xi]*px[xi]
    return pym_c


def p_m(pmx_c,px):
    """Update marginal distribution of bottleneck variable.
    
    Arguments:
    pmx_c -- conditional distribution p(M|X)
    px -- marginal distribution p(X)
    
    Returns:
    pm - marginal distribution of compression p(M)
    """
    pm = np.zeros(pmx_c.shape[0])
    for mi in range(pm.size):
        for xi in range(px.size):
            pm[mi] += pmx_c[mi,xi]*px[xi]
    return pm


embo/core.py
# Copyright (C) 2020,2021 Eugenio Piasini, Alexandre Filipowicz,
# Jonathan Levine.
#
# This file is part of Embo.
#
# Embo is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Embo is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
# License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Embo.  If not, see <https://www.gnu.org/licenses/>.

import numpy as np
import multiprocessing as mp

from .utils import p_joint, mi_x1x2_c, compute_upper_bound, kl_divergence, entropy

np.seterr(divide='ignore', invalid='ignore')

class InformationBottleneck:

    def __init__(self, x=None, y=None, alpha=1, window_size_x=1, window_size_y=1, pxy=None, **kwargs):
        """ Information Bottleneck analysis for an empirical dataset (x,y) or a joint probability mass function pxy.

            Arguments:
            x -- first empirical observation sequence ("past" if doing past-future bottleneck analysis)
            y -- second empirical observation sequence ("future")
            alpha -- generalized bottleneck parameter: alpha=1 is IB and alpha=0 is DIB (deterministic bottleneck)
            window_size_x, window_size_y (int) -- size of the moving windows to be used to compute the IB curve (you typically don't need to worry about this unless you're doing a "past-future bottleneck"-type analysis). The time window on x (which in these cases is typically the "past") is taken backwards, and the time window on y (the "future") is taken forwards. For instance, setting window_size_x=3 and window_size_y=2 will yield the IB curve between (X_{t-2},X_{t-1},X_{t}) and (Y_{t},Y_{t+1}).
            pxy -- joint probability distribution of X and Y. This is a numpy array such that pxy[xi,yi] is the joint probability of the xi-th value of X and the yi-th value of Y. If the array does not sum to one, it will be normalized to ensure that it does. If this argument is passed, the joint estimation step will be skipped; therefore it can only be passed if x and y are None.
            kwargs -- additional keyword arguments to be passed to IB().

        """

        self.x = np.asarray(x)
        self.y = np.asarray(y)
        self.alpha = alpha
        self.window_size_x = window_size_x
        self.window_size_y = window_size_y
        self.kwargs_IB = kwargs
        self.results_ready = False
        if (x is not None and not np.all(np.isfinite(self.x))) or\
           (y is not None and not np.all(np.isfinite(self.y))):
            raise ValueError("The observation data contains NaNs or Infs.")
        if pxy is not None and (x is not None or y is not None):
            raise ValueError("It is not possible to specify both pxy and the empirical data x,y.")
        self.pxy_j = pxy
        if self.pxy_j is not None:
            if not np.all(self.pxy_j>=0):
                raise ValueError("Negative values in the specified joint p(X,Y).")
            self.pxy_j /= 1.0*self.pxy_j.sum()
        elif x is None or y is None:
            raise ValueError("Eiter pxy or x and y should be specified.")
        elif len(x)==0 or len(y)==0:
            raise ValueError("If pxy is not specified, x and y can't be empty.")
        self.get_empirical_bottleneck = self.get_bottleneck

    def compute_IB_curve(self):
        """ Compute the IB curve for the joint empirical observations for X and Y. """
        
        # Marginal, joint and conditional distributions required to calculate the IB
        if self.pxy_j is None:
            self.pxy_j = p_joint(self.x, self.y, self.window_size_x, self.window_size_y)

        px = self.pxy_j.sum(axis=1)
        py = self.pxy_j.sum(axis=0)
        pyx_c = self.pxy_j.T / px

        # Calculate the information bottleneck for a range of values of beta
        self.i_x, self.i_y, self.h_m, self.beta, self.mixy, self.hx = self.IB(px, py, pyx_c, self.alpha, **self.kwargs_IB)
        self.hy = entropy(py)

        # set a flag we will use to call this function automatically when needed
        self.results_ready = True
    
    def get_bottleneck(self, return_entropies=False):
        """Return array of I(M:X) and I(M:Y) for array of different values of beta

         Returns:
            i_x -- values of I(M:X) for each value of beta
            i_y -- values of I(M:Y) for each value of beta
            h_m -- values of H(M) for each value of beta
            beta -- values of beta considered
            mixy -- mutual information between X and Y, I(X:Y) (curve saturation point) (only returned if return_entropies is True)
            hx -- entropy of X (only returned if return_entropies is True)
            hy -- entropy of Y (only returned if return_entropies is True)
        """
        if not self.results_ready:
            self.compute_IB_curve()
        
        if return_entropies:
            return self.i_x, self.i_y, self.h_m, self.beta, self.mixy, self.hx, self.hy
        else:
            return self.i_x, self.i_y, self.h_m, self.beta
    
    def get_ix(self):
        if not self.results_ready:
            self.compute_IB_curve()
        return self.i_x

    def get_iy(self):
        if not self.results_ready:
            self.compute_IB_curve()
        return self.i_y

    def get_hm(self):
        if not self.results_ready:
            self.compute_IB_curve()
        return self.h_m

    def get_beta_values(self):
        if not self.results_ready:
            self.compute_IB_curve()
        return self.beta

    def get_saturation_point(self):
        if not self.results_ready:
            self.compute_IB_curve()
        return self.mixy

    def get_entropies(self):
        if not self.results_ready:
            self.compute_IB_curve()
        return self.hx, self.hy

    @classmethod
    def beta_iter(cls, a, b, px, py, pyx_c, pm_size, restarts, iterations, rtol=1e-3, atol=0):
        """Function to run BA algorithm for individual values of beta

        Arguments:
        a -- value of alpha defining the generalized bottleneck (a=1 is IB, a=0 is DIB)
        b -- value of beta on which to run algorithm
        px -- marginal probability distribution for X
        py -- marginal probability distribution for Y
        pyx_c -- conditional distribution p(y|x)
        pm_size -- discrete size of the compression distribution
        restarts -- number of times the optimization procedure should be restarted (for each value of beta) from different random initial conditions
        iterations -- maximum number of iterations to perform
        rtol, atol -- relative and absolute tolerances on the cost function to determine when the algorithm has converged

        Returns:
        list with i_x and i_y, which correspond to I(M:X) and I(M:Y) values for each value of beta
        """
        candidates = []
        for r in range(restarts):

            # initialization
            if pm_size==px.size and pm_size>0:
                # by default pm_size will be the same as px.size. In
                # this case we initialize similarly to Strouse and
                # Schwab 2016: for x_j, most of the mass of the
                # encoder p(m|x) is on m_j, while a smaller part of
                # the mass is distributed randomly over the other
                # values of m.
                pmx_c = np.zeros((pm_size, px.size))
                delta = 0.2 * np.random.rand(1,px.size)
                pmx_c[0,:] = 0.75 - delta
                pmx_c[1:,:] = np.random.rand(pm_size-1, px.size)+1
                pmx_c[1:,:] = (0.25+delta)*pmx_c[1:,:]/(pmx_c[1:,:].sum(axis=0))
                for kx in range(pmx_c.shape[1]):
                    pmx_c[:,kx] = np.roll(pmx_c[:,kx], kx)
                pm = p_m(pmx_c, px)
                pym_c = p_ym_c(pm, px, pyx_c, pmx_c)
            else:
                # legacy initialization scheme
                pm = np.random.rand(pm_size)+1
                pm /= pm.sum()
                pym_c = np.random.rand(py.size, pm.size)+1  # Starting point for the algorithm
                pym_c /= pym_c.sum(axis=0)
                    
            
            # Iterate the BA algorithm
            for i in range(iterations):
                pmx_c, z = p_mx_c(pm, px, pyx_c, pym_c, a, b)
                pm = p_m(pmx_c,px)
                pm, pmx_c, pym_c = drop_unused_dimensions(pm, pmx_c, pym_c)
                pym_c = p_ym_c(pm, px, pyx_c, pmx_c)
                pm, pmx_c, pym_c = drop_unused_dimensions(pm, pmx_c, pym_c)
                # compute cost: H(M)-αH(M|X)-βI(M:Y)
                if a==1:
                    # specialized efficient form for the regular IΒ
                    # (eq. 29 in Tishby 2000; equivalent to the
                    # formula below, but faster)
                    cost = -px @ np.log(z)
                else:
                    # generalized bottleneck
                    cost = entropy(pm) - a * px @ entropy(pmx_c, axis=0) - b * mi_x1x2_c(py, pm, pym_c)
                if i > 0 and np.allclose(cost, cost_old, rtol=rtol, atol=atol):
                    # if the cost function is not changing any more, we're at convergence and we can stop
                    break
                cost_old = cost
            candidates.append({'info_x': mi_x1x2_c(pm, px, pmx_c),
                               'info_y': mi_x1x2_c(py, pm, pym_c),
                               'entropy_m' : entropy(pm),
                               'cost': cost})
        # among the restarts, select the result with minimum cost
        selected_candidate = min(candidates, key=lambda c: c['cost'])
        i_x = selected_candidate['info_x']
        i_y = selected_candidate['info_y']
        h_m = selected_candidate['entropy_m']
        return [i_x, i_y, h_m, b]

    @classmethod
    def IB(cls, px, py, pyx_c, alpha=1, minsize = False, maxbeta=5, minbeta=1e-2, numbeta=30, iterations=100, restarts=3, processes=1, ensure_monotonic_bound='auto', rtol=1e-3):
        """Compute an Information Bottleneck curve

        Arguments:
        px -- marginal probability distribution for X
        py -- marginal probability distribution for Y
        pyx_c -- conditional probability of Y given X
        alpha -- generalized bottleneck parameter: alpha=1 is IB, alpha=0 is DIB
        minsize -- if True, maximum size of compression matches smallest codebook size between x and y
        maxbeta -- the maximum value of beta to use to compute the curve.
        minbeta -- the minimum value of beta to use.
        numbeta -- the number of (equally-spaced) beta values to consider to compute the curve.
        iterations -- number of iterations to use to for the curve to converge for each value of beta
        restarts -- number of times the optimization procedure should be restarted (for each value of beta) from different random initial conditions.
        processes -- number of cpu threads to run in parallel (default = 1)
        ensure_monotonic_bound -- one of 'entropy', 'information', 'auto' or False. This parameter can be used to ensure that the final bound is monotonic in the chosen plane. If 'information', monotonicity is encforced in the IB plane. If 'entropy', monotonicity is enforced in the DIB plane. If 'auto', monotonicity is enforced in the IB plane for alpha=1 and in the DIB plane for alpha==0, but not for intermediate values. If False, monotonicity is not enforced and the original beta sequence is preserved.
        rtol -- relative tolerance parameter for convergence criterion.

        Returns:
        ips -- values of I(M:X) for each beta considered
        ifs -- values of I(M:Y) for each beta considered
        hms -- values of H(M) for each beta considered
        bs -- values of beta considered. Note that this does not necessarily match the set of values initially considered (numbeta equally spaced values between 0.01 and maxbeta). As numerical accuracies can lead to to solutions (I(M:X), I(M:Y)) that would make the IB curve nonmonotonic, any such solution (as well as the corresponding beta value) is discarded before returning the results. See utils.compute_upper_bound() for more details.
        mixy -- mutual information between x and y (curve saturation point)
        hx -- entropy of x (maximum I(M:X) value)
        """
        pm_size = px.size
        if minsize:
            pm_size = min(px.size,py.size)  # Get compression size - smallest size
        
        bs = np.linspace(minbeta, maxbeta, numbeta)  # value of beta

        # Parallel computing of compression for desired beta values
        with mp.Pool(processes=processes) as pool:
            results = [pool.apply_async(cls.beta_iter, args=(alpha, b, px, py, pyx_c, pm_size, restarts, iterations,rtol)) for b in bs]
            results = [p.get() for p in results]
        ips = [x[0] for x in results]
        ifs = [x[1] for x in results]
        hms = [x[2] for x in results]
        bs = [x[3] for x in results]

        # Values of beta may not be sorted appropriately due to out-of
        # order execution if using many processes. So we have so sort
        # the result lists (ix, iy, hm, beta) in ascending beta order.
        ips = [x for _, x in sorted(zip(bs, ips))]
        ifs = [x for _, x in sorted(zip(bs, ifs))]
        hms = [x for _, x in sorted(zip(bs, hms))]
        bs = sorted(bs)

        # restrict the returned values to those that, at each value of
        # beta, actually increase (for I(M:X) or H(M)) and do not
        # decrease (for I(M:Y)) the information/entropy with respect
        # to the previous value of beta. This is to avoid confounds
        # from cases where the AB algorithm gets stuck in a local
        # minimum.
        if ensure_monotonic_bound:
            if ensure_monotonic_bound=='auto':
                if alpha==1:
                    # If alpha ==1 (vanilla IB), we ensure monotonicity in the IB plane
                    ensure_monotonic_bound = 'information'
                elif alpha==0:
                    # If alpha==0, we ensure monotonicity in the DIB plane
                    ensure_monotonic_bound = 'entropy'
            if ensure_monotonic_bound=='information':
                ub, idxs = compute_upper_bound(ips, ifs)
                ips = np.squeeze(ub[:, 0])
                ifs = np.squeeze(ub[:, 1])
                hms = np.array(hms)[idxs]
                bs = np.array(bs)[idxs]
            elif ensure_monotonic_bound=='entropy':
                ub, idxs = compute_upper_bound(hms, ifs)
                hms = np.squeeze(ub[:, 0])
                ifs = np.squeeze(ub[:, 1])
                ips = np.array(ips)[idxs]
                bs = np.array(bs)[idxs]

        # Return saturation point (mixy) and max horizontal axis (hx)
        mixy = mi_x1x2_c(py, px, pyx_c)
        hx = entropy(px)
        return ips, ifs, hms, bs, mixy, hx

def elementwise_l(pm, px, pyx_c, pym_c, alpha, beta):
    """Log-loss function, elementwise.

    This is the generalization of l_β(x,m) in Strouse 2016 to the case
    where 0<α≤1.

    """
    return (np.log(pm[:,np.newaxis]) - beta * kl_divergence(pyx_c[:,np.newaxis,:], pym_c[:,:,np.newaxis]))/alpha

def p_mx_c(pm, px, pyx_c, pym_c, alpha, beta):
    """Update conditional distribution of bottleneck random variable given x.

    Arguments:
    pm -- marginal distribution p(M) - vector
    px -- marginal distribution p(X) - vector
    pyx_c -- conditional distribution p(Y|X) - matrix
    pym_c -- conditional distribution p(Y|M) - matrix
    alpha -- H(M|X) weight: alpha=1 is IB, alpha=0 is DIB
    beta -- I(M:Y) weight

    Returns:
    pmx_c -- conditional distribution p(M|X) (encoder)
    z -- normalizing factor
    """
    if alpha > 0:
        # Generalized Information Bottleneck. Note that vanilla IB
        # corresponds to the alpha=1 case.
        pmx_c = np.exp(elementwise_l(pm, px, pyx_c, pym_c, alpha, beta))
        z = pmx_c.sum(axis=0)
        pmx_c /= z # normalize
    elif alpha==0:
        # Deterministic Information Bottleneck. As per Algorithm 2 in
        # Strouse 2016, we compute the log-loss function for the
        # vanilla IB (α=1), and for each value of x we set it to 1 for
        # the value of t for which it's maximum, and zero otherwise.
        l = elementwise_l(pm, px, pyx_c, pym_c, 1, beta)
        pmx_c = np.zeros_like(l)
        pmx_c[np.argmax(l, axis=0), np.arange(l.shape[1])] = 1
        z = 1
    return pmx_c, z

def p_ym_c(pm, px, pyx_c, pmx_c):
    """Update conditional distribution of bottleneck variable given y.

    Arguments:
    pm -- Marginal distribution p(M)
    px -- marginal distribution p(X)
    pyx_c -- conditional distribution p(Y|X)
    pmx_c -- conditional distribution p(M|X)

    Returns:
    pym_c -- conditional distribution p(Y|M)
    """
    pym = pyx_c*px[np.newaxis,:] @ pmx_c.T
    pym_c = pym / pm[np.newaxis,:]
    return pym_c

def p_m(pmx_c, px):
    """Update marginal distribution of bottleneck variable.

    Arguments:
    pmx_c -- conditional distribution p(M|X)
    px -- marginal distribution p(X)

    Returns:
    pm - marginal distribution of compression p(M)
    """
    return pmx_c @ px

def drop_unused_dimensions(pm, pmx_c, pym_c, eps=0):
    """Remove bottleneck dimensions that are not in use anymore.

    This is in general safe to do as once a dimension is excluded by
    the algorithm, it is never brought back. It is necessary to use
    this function particularly when using the generalized bottleneck,
    where dimensions are discarded very aggressively, and if keeping
    them we would need to put special safeguards in place to avoid
    NaNs and infinities in the Dkl and when computing p(y|m).

    """
    unused_pm = pm<=eps
    unused_pmx_c = np.all(pmx_c<=eps, axis=1)
    unused_pym_c = np.all(pym_c<=eps, axis=0)
    unused = np.any(np.vstack([unused_pm, unused_pmx_c.T, unused_pym_c]), axis=0)
    in_use = ~unused
    return pm[in_use], pmx_c[in_use,:], pym_c[:,in_use]
